src/order/heyting/regular.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.galois_connection
+
+/-!
+# Heyting regular elements
+
+This file defines Heyting regular elements, elements of an Heyting algebra that are their own double
+complement, and proves that they form a boolean algebra.
+
+From a logic standpoint, this means that we can perform classical logic within intuitionistic logic
+by simply double-negating all propositions. This is practical for synthetic computability theory.
+
+## Main declarations
+
+* `is_heyting_regular`: `a` is Heyting-regular if `aᶜᶜ = a`.
+* `heyting_regular`: The subtype of Heyting-regular elements.
+* `heyting_regular.boolean_algebra`: Heyting-regular elements form a boolean algebra.
+
+## References
+
+* [Francis Borceux, *Handbook of Categorical Algebra III*][borceux-vol3]
+-/
+
+open function
+
+variables {α : Type*}
+
+section heyting_algebra
+variables [heyting_algebra α] {a b : α}
+
+/-- An element of an Heyting algebra is regular if its double complement is itself. -/
+def is_heyting_regular (a : α) : Prop := aᶜᶜ = a
```

Comment:
An alternative would be to define `heyting_regular` as a `set α` and replace `is_heyting_regular x` with `x ∈ heyting_regular α`, any reason that you went for this definition instead?

Reply:
Dot notation is a convincing enough reason (thinking of all the annoyances with `non_zero_divisors` not having dot notation.)